pilot/pkg/serviceregistry/kube/controller/ambient/workloads.go

```diff
@@ -0,0 +1,396 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// nolint: gocritic
+package ambient
+
+import (
+	"net/netip"
+
+	v1 "k8s.io/api/core/v1"
+
+	networkingv1alpha3 "istio.io/api/networking/v1alpha3"
+	networkingclient "istio.io/client-go/pkg/apis/networking/v1alpha3"
+	securityclient "istio.io/client-go/pkg/apis/security/v1beta1"
+	"istio.io/istio/pilot/pkg/features"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pkg/config/constants"
+	"istio.io/istio/pkg/config/labels"
+	"istio.io/istio/pkg/config/schema/kind"
+	kubeutil "istio.io/istio/pkg/kube"
+	"istio.io/istio/pkg/kube/krt"
+	kubelabels "istio.io/istio/pkg/kube/labels"
+	"istio.io/istio/pkg/log"
+	"istio.io/istio/pkg/ptr"
+	"istio.io/istio/pkg/slices"
+	"istio.io/istio/pkg/spiffe"
+	"istio.io/istio/pkg/workloadapi"
+)
+
+func (a *index) WorkloadsCollection(
+	Pods krt.Collection[*v1.Pod],
+	MeshConfig krt.Singleton[MeshConfig],
+	AuthorizationPolicies krt.Collection[model.WorkloadAuthorization],
+	PeerAuths krt.Collection[*securityclient.PeerAuthentication],
+	Waypoints krt.Collection[Waypoint],
+	WorkloadServices krt.Collection[model.ServiceInfo],
+	WorkloadEntries krt.Collection[*networkingclient.WorkloadEntry],
+	ServiceEntries krt.Collection[*networkingclient.ServiceEntry],
+	AllPolicies krt.Collection[model.WorkloadAuthorization],
+) krt.Collection[model.WorkloadInfo] {
```

Comment:
Echoing @stevenctl a bit here - this is still effectively a series of very large, behaviorally-critical funcs that are somewhat intricate, and not unit-tested _here_ AFAICT (I am sure we are indirectly testing this with integ tests elsewhere). 

That makes this section of code a bit more intimidating to work on and validate in isolation than it needs to be.

I would like to see unit test coverage/test-based encoding of expected behavior of e.g. the Workload construction.

Reply:
We do have almost full coverage  via the ambientindex_tests.go. I can look into how we can isolate things to make them completely trivial stateless tests maybe